packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -4089,6 +4091,64 @@ class _DeleteTextAction<T extends DirectionalTextEditingIntent> extends ContextA
   bool get isActionEnabled => !state.widget.readOnly && state._value.selection.isValid;
 }
 
+/// An [Action] that transposes the characters immediately before and after the
+/// current collapsed selection.
+///
+/// When the cursor is at the end of the text, transposes the last two
+/// characters, if they exist.
+///
+/// When the cursor is at the start of the text, does nothing.
+class _TransposeCharactersAction extends ContextAction<TransposeCharactersIntent> {
+  _TransposeCharactersAction(this.state);
+
+  final EditableTextState state;
+
+  @override
+  Object? invoke(TransposeCharactersIntent intent, [BuildContext? context]) {
+    // Can't swap 1 or 0 characters.
+    if (state.textEditingValue.text.length <= 1) {
```

Comment:
maybe count characters instead of code units?

Reply:
Great catch!